### ISSUE TITLE
LoRA adapters weights remain zero after training with deepspeed

### ISSUE THREAD

**Issue by schwarzwalder93** [huggingface/peft#1402]
### System Info

torch 2.0.1
transformers 4.37.0
accelerate 0.26.1
deepspeed 0.12.4
peft 0.7.1

### Who can help?

@pacman100 @ArthurZucker @VictorSanh 

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

When using `transformers.Trainer` and `TrainingArguments` with `deepspeed=ds_config`, I noticed that the LoRA adapter weights (`lora_B`) remain zero even after the training is over. However, the training seems to work smoothly and the train/val loss have significantly reduced. But since the `lora_B` matrix remains zero, the PEFT model is same as base model.

Code Snippet - https://colab.research.google.com/drive/1sRb56JMxXUang_VH_iLCZpt8gAeBR8Wp?usp=sharing

When I comment out `deepspeed=ds_config` from `TrainingArguments` and run the training again, the `lora_B` weights have non-zero float values as expected. 

### Expected behavior

Expected `lora_B` weights to be non-zeros, when using deepspeed for training.

**Comment by VictorSanh** [comment]
@leot13 could you look into this? wondering if it's not related to your problem you have been working on

**Comment by mehrdad-es** [comment]
@schwarzwalder93, how do the train/val loss look like when you have commented out `deepspeed=ds_config`?

**Comment by YiFraternity** [comment]
I encountered the same problem.

System info

torch 2.0.1
transformers 4.33.1
accelerate  0.26.1
deepspeed 0.9.4
cuda 11.7
peft 0.7.1

**Comment by leot13** [comment]
> @leot13 could you look into this? wondering if it's not related to your problem you have been working on

Yes it's probably related. This is most likely an issue due to using both peft and gradient checkpointing. Basically, when using gradient checkpointing with peft, you have to then attach hooks requiring grads to the output of the first layer. It's done [here](https://github.com/huggingface/transformers/blob/f1cc6157210abd9ea953da097fb704a1dd778638/src/transformers/modeling_utils.py#L2142), by using the function [enable_gradient_checkpointing()](https://github.com/huggingface/transformers/blob/f1cc6157210abd9ea953da097fb704a1dd778638/src/transformers/modeling_utils.py#L1576). 
The issue with models like idefics is that the [get_input_embeddings](https://github.com/huggingface/transformers/blob/f1cc6157210abd9ea953da097fb704a1dd778638/src/transformers/modeling_utils.py#L1593) will get you the embedding layer of the language model component, but not the vision one. 
You should be able to fix this by redefining enable_gradient_checkpointing() with something that looks like this: 
```
def enable_input_require_grads(self):
        """
        Enables the gradients for the input embeddings. This is useful for fine-tuning adapter weights while keeping
        the model weights fixed.
        """

        def get_lowest_module(module):
            if len(list(module.children())) == 0:
                # If the module has no children, it is a leaf module (e.g., Linear, Conv2d, etc.)
                return module
            else:
                # Recursively call the function on each child module
                return get_lowest_module(list(module.children())[0])

        def make_inputs_require_grads(module, input, output):
            output.requires_grad_(True)

        self._text_require_grads_hook = self.get_input_embeddings().register_forward_hook(make_inputs_require_grads)
        self._vision_require_grads_hook = get_lowest_module(self.vision_model).register_forward_hook(
            make_inputs_require_grads
        )
```

Alternatively, if you don't need it, you can also just set gradient_checkpointing to False in the trainer

**Comment by YiFraternity** [comment]
> I encountered the same problem.
> 
> System info
> 
> torch 2.0.1 transformers 4.33.1 accelerate 0.26.1 deepspeed 0.9.4 cuda 11.7 peft 0.7.1

问题在于model.save_pretrained，不能保存训练后的B矩阵

**Comment by YiFraternity** [comment]
I know the reason now, it's because you also used the Alpaca Lora framework. You used the following code:
```py
    model.config.use_cache = False

    old_state_dict = model.state_dict
    model.state_dict = (
        lambda self, *_, **__: get_peft_model_state_dict(
            self, old_state_dict()
        )
    ).__get__(model, type(model))
```
Simply deleting the aforementioned code will suffice, enabling the preservation of the B matrix.

**Comment by schwarzwalder93** [comment]
@leot13 Thanks for the hook. That works!

**Comment by Vaibhavsun** [comment]
Hi i have too this problem when using Peft for BloomSequenceClassification my lora_A adapter weights becoming non zero while lora_B weights becoming zero what to do and my gradient_checkpointing is also False

`import` modal
from pathlib import Path

import modal.gpu

pth=Path("/Models")
app=modal.App("FFL")
volume=modal.Volume.from_name("FFL",create_if_missing=True)
image=modal.Image.debian_slim(python_version="3.10").apt_install("libgl1","libglib2.0-0").pip_install(["langchain-core==0.3.15","langchain-google-vertexai","langchainhub","faiss-cpu","langchain_community","langchain-huggingface","sentence-transformers==2.6.0","huggingface_hub","torch","transformers==4.45.2","numpy","pandas","torchvision","Pillow","scikit-learn","sentencepiece","peft","bitsandbytes","datasets"])
@app.function(image=image,gpu=modal.gpu.A10G(count=1),timeout=86400,volumes={pth:volume})
def main():
    import os
    import numpy as np

# Set CUDA_VISIBLE_DEVICES to limit to GPU 0
    os.environ['CUDA_VISIBLE_DEVICES'] = '0'

    import torch
    from transformers import BloomForSequenceClassification, BloomTokenizerFast, TrainingArguments, Trainer
    from datasets import load_dataset
    from peft import get_peft_model, LoraConfig, TaskType
    from sklearn.metrics import accuracy_score
    from transformers import DataCollatorWithPadding

    # Check CUDA availability and set device
    if torch.cuda.is_available():
        # torch.cuda.set_device(3)  # Explicitly set to use GPU 0
        device = torch.device('cuda')
        print(f"Using CUDA device {torch.cuda.current_device()}: {torch.cuda.get_device_name(0)}")
    else:
        device = torch.device('cpu')
        print("CUDA is not available. Using CPU.")

    # Print CUDA information
    if torch.cuda.is_available():
        print(f"CUDA version: {torch.version.cuda}")
        print(f"Number of available GPUs: {torch.cuda.device_count()}")
        print(f"Current GPU memory usage: {torch.cuda.memory_allocated(0) / 1024**2:.2f} MB")

    # Preprocess function
    def preprocess_data(example):
        encoded = tokenizer(example['question'], example['sentence'], truncation=True, padding='max_length', max_length=512)
        return {
            'input_ids': encoded['input_ids'],
            'attention_mask': encoded['attention_mask'],
            'labels': example['label']
    }
    # Partition data using Dirichlet distribution
    def partition_data(dataset, num_clients, alpha=0.3, seed=0):
        np.random.seed(seed)
        labels = [example['labels'] for example in dataset]
        num_labels = len(set(labels))
        label_indices = {label: [] for label in set(labels)}
        
        for i, label in enumerate(labels):
            label_indices[label].append(i)
        
        dirichlet_weights = np.random.dirichlet([alpha] * num_labels, size=num_clients)
        
        client_data = [[] for _ in range(num_clients)]
        
        for label, indices in label_indices.items():
            num_indices = len(indices)
            proportions = dirichlet_weights[:, list(label_indices.keys()).index(label)]
            allocations = np.round(proportions * num_indices).astype(int)
            
            for client_idx in range(num_clients):
                client_data[client_idx].extend(np.random.choice(indices, size=allocations[client_idx], replace=False))
        
        client_data = [dataset.select(indices) for indices in client_data]
        return client_data
    
    import torch
    def federated_average(global_model, client_models, client_sizes):
        total_size = sum(client_sizes)
        global_dict = global_model.state_dict()
        
        for k in global_dict.keys():
            if 'lora' in k:  # Only update LoRA parameters
                global_dict[k] = torch.stack([
                    client_models[i].state_dict()[k] * (client_sizes[i] / total_size)
                    for i in range(len(client_models))
                ]).sum(dim=0)
        
        global_model.save_pretrained(pth/"bloom_peft_fedavg_model")
        torch.save(global_model.state_dict(),pth/'model_weights_round1.pth')
        return global_model
    from peft import PeftModel
    tokenizer = BloomTokenizerFast.from_pretrained('bigscience/bloom-560m')
    model_name = 'bigscience/bloom-560m'
    model = BloomForSequenceClassification.from_pretrained(model_name, num_labels=2).to(device)
    peft_config = LoraConfig(
    task_type=TaskType.SEQ_CLS,
    r=16,
    lora_alpha=8,
    lora_dropout=0.1,
    bias="all",
    target_modules=["query_key_value"]
)
    model = get_peft_model(model, peft_config)
    model.print_trainable_parameters()



# # Load and preprocess dataset
    dataset = load_dataset("glue", "qnli")['train'].select(range(10000))
    preprocessed_dataset = dataset.map(preprocess_data, remove_columns=dataset.column_names)
    print("Dataset Created")
    # Partition dataset among clients using Dirichlet distribution
    num_clients = 20
    client_datasets = partition_data(preprocessed_dataset, num_clients)








    # Define compute metrics function









    def compute_metrics(eval_pred):
        predictions = np.argmax(eval_pred.predictions, axis=1)
        return {"accuracy": accuracy_score(eval_pred.label_ids, predictions)}

    # Federated Learning parameters
    local_epochs = 2
    num_rounds = 15 #21
    clients_per_round = 4

    # Define training arguments with GPU 0 specific settings
    training_args = TrainingArguments(
        output_dir="./results",
        learning_rate=5e-5,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        num_train_epochs=local_epochs,
        evaluation_strategy="no",
        save_strategy="no",
        fp16=torch.cuda.is_available(),
        gradient_accumulation_steps=4,
        dataloader_pin_memory=False,
        optim="adamw_torch",
        no_cuda=not torch.cuda.is_available(),
        lr_scheduler_type="cosine",  # Options: "linear", "cosine", "cosine_with_restarts", etc.
    )

    data_collator = DataCollatorWithPadding(tokenizer, padding=True, return_tensors="pt")
    # Initialize federated learning rounds
    # model=PeftModel.from_pretrained(model,'./bloom_peft_fedavg_model',config=peft_config,is_trainable=True)
    # model.load_state_dict(torch.load('model_weights.pth'))
    global_model = model
    from transformers import TrainerCallback
    import torch

    class LoRAGradientTracker(TrainerCallback):
        def on_step_end(self, args, state, control, model, **kwargs):
            print("\nTracking LoRA Gradients:")
            for name, param in model.named_parameters():
                if param.grad==None:
                    continue
                grad_values = param.grad.detach().cpu().numpy()
                if ("lora_A" in name):
                    print(f"Gradients for {name}: {grad_values}")
                if ("lora_b" in name):
                    print(f"Gradients for {name}: {grad_values}")


    for round_num in range(num_rounds):
        print(f"Round {round_num + 1}/{num_rounds}")
        
        # Sample clients
        sampled_clients = np.random.choice(range(num_clients), size=clients_per_round, replace=False)
        client_models = []
        client_sizes = []
        for client_id in sampled_clients:
            print(f"Training client {client_id}")
            
            # Create a new base model
            base_model = BloomForSequenceClassification.from_pretrained(model_name, num_labels=2)
            
            # Create a new PEFT model
            client_model = get_peft_model(base_model, peft_config)

            # client_model=global_model
            # Load the state dict from the global model
            client_model.load_state_dict(global_model.state_dict())
            client_model.to(device)
            
            client_dataset = client_datasets[client_id]
            client_sizes.append(len(client_dataset))
            print(f"Current GPU memory usage: {torch.cuda.memory_allocated(0) / 1024**2:.2f} MB")

            trainer = Trainer(
                model=client_model,
                args=training_args,
                train_dataset=client_dataset,
                data_collator=data_collator,
                compute_metrics=compute_metrics,
            )
            trainer.train()
            client_model.to('cpu')
            client_models.append(client_model)
            print('lora_b')
            for keys in client_model.state_dict():
                if 'lora_B' in keys:
                    print(global_model.state_dict()[keys])
            print('lora_A')
            for keys in client_model.state_dict():
                if 'lora_A' in keys:
                    print(global_model.state_dict()[keys])
            del client_model,trainer,base_model
        
        # Update global model using FedAvg
        global_model = federated_average(global_model, client_models, client_sizes)

        for keys in global_model.state_dict():
            if 'lora_B' in keys:
                print(global_model.state_dict()[keys])

        print("Lora a VALUES")
        for keys in global_model.state_dict():
            if 'lora_A' in keys:
                print(global_model.state_dict()[keys])
        

                                        
            
                                        
                                        
                                        
                                        

    # Save global model
    global_model.save_pretrained("./bloom_peft_fedavg_model")

    print("Federated Learning completed successfully!")
# Configure LoRA
    for keys in global_model.state_dict():
        if 'lora_B' in keys:
            print(global_model.state_dict()[keys])

    print("Lora a VALUES")
    for keys in global_model.state_dict():
        if 'lora_A' in keys:
            print(global_model.state_dict()[keys])
